### ISSUE TITLE
System.Reflection.Emit.AssemblyBuilder.Save

### ISSUE THREAD

**Issue by jonorossi** [dotnet/corefx#4491]
`AssemblyBuilder.Save` and `AssemblyBuilderAccess.RunAndSave` isn't available in .NET Core, however coreclr seems to have the code to implement it but I looks conditionally compiled out.

https://github.com/dotnet/coreclr/blob/bc146608854d1db9cdbcc0b08029a87754e12b49/src/mscorlib/src/System/Reflection/Emit/AssemblyBuilder.cs#L1690-L1711

Our use case in Castle DynamicProxy is to write out dynamically created assemblies to disk so we can run peverify over the assembly in unit tests.

Is there any chance we can get this functionality that .NET Framework has introduced into .NET Core, or an alternative way of doing the same thing. This might not sound like a big deal if you aren't familiar with Castle DynamicProxy's internals, but it really is as it heavily reduces our confidence in both our code and the .NET runtime as well as reducing our ability to track down defects.

**Comment by joshfree** [comment]
Related issue https://github.com/dotnet/coreclr/issues/1709

**Comment by bartdesmet** [comment]
Another interesting omission is the lack of LambdaExpression.Compile accepting a MethodBuilder which tends to be very useful when combined with AssemblyBuilder.Save support.

**Comment by jkotas** [comment]
Large chunks of Reflection.Emit code in the runtime are Windows-specific. It is not easy to just enable them for CoreCLR because of it would not work on Unix. In particular, the support for emitting debug information depends on unmanaged PDB writer that is very complex Windows-specific component that we have no plans to open source and bring x-plat.

We are also not happy about Reflection.Emit being integrated deeply into the runtime. Reflection.Emit should be a independent higher level component that generates PE file that the runtime loads as any other assembly (note that runtime supports loading assemblies from byte array, so the PE file does not have to be persisted on disk). IKVM.Reflection.Emit done by Jeroen Frijters proved that it is pretty doable and can actually work much better than the current Reflection.Emit built into the runtime.

The plan to address the Reflection.Emit is:

1. Create standalone managed metadata, IL and portable PDB writer that can be used by Roslyn, and anybody else to produce managed assemblies. @tmat has been working it: https://github.com/dotnet/roslyn/pull/7683.

2. Build Reflection.Emit object model as a layer on top of the standalone managed metadata writer to make it easier for existing projects to migrate to .NET Core.

**Comment by jonorossi** [comment]
@jkotas many thanks for the explanation here, it makes complete sense. I'm glad you guys are planning to provide a more general purpose object model around the heart of .NET, IL. I'd seen the portable PDB reader/writer and some metadata object model appear in Roslyn, they look great. In the past I've used Cecil for the exact reason the .NET Framework API is so heavily tied into the runtime, glad to see more stuff being pulled out of the runtime.

**Comment by tmat** [comment]
FYI @ManishJayaswal @jaredpar

**Comment by karelz** [comment]
Next step: We need API proposal with deep analysis how implementable it is. The old Desktop API might be an option, likely not a good one.
This is quite hard to do and requires lots of experience.

**Comment by KevinRansom** [comment]
@karelz @jkotas --- What is the status of this feature?  When can we expect to see dynamic codegen with persistence on the coreclr?

**Comment by karelz** [comment]
There is currently no plan (milestone Future + my comment above) - area owners @AtsushiKan @DnlHarvey @joshfree can provide more details.

**Comment by stakx** [comment]
@jkotas:

> We are also not happy about Reflection.Emit being integrated deeply into the runtime. Reflection.Emit should be a independent higher level component that generates PE file that the runtime loads as any other assembly [...].
>
> The plan to address the Reflection.Emit is:
>
> 1. [...]
> 2. Build Reflection.Emit object model as a layer on top of the standalone managed metadata writer to make it easier for existing projects that use Reflection.Emit to migrate to .NET Core.

I do not fully understand the plan, in particular with regard to dynamic assemblies (`AssemblyBuilder.DefineDynamicAssembly`) and their ability to be incrementally updated / augmented with new dynamic types. How would this still be possible if assemblies must first be written out as a byte stream or file, then be loaded?

I imagine one would have to write each dynamically generated type to a single assembly... but wouldn't this interfere with type visibility and type member accessibility between generated types, and with stuff like `[assembly: InternalsVisibleTo("DynamicallyGeneratedAssembly")`?

**Comment by jkotas** [comment]
> their ability to be incrementally updated / augmented with new dynamic types

Yes, there would need to be a runtime hook for this.

Conceptually, each type addition in Reflection.Emit produces a new assembly that contains all existing metadata (with all existing tokens preserved) plus the new type. Then the runtime is asked to update the metadata of the already loaded assembly to the new version.

Sending the whole assembly over for each edit is not efficient for large assemblies with small edits. It would be better to just send the metadata delta with the new stuff over. Fortunately, this exists in the runtime for edit and continue already. Look for https://github.com/dotnet/coreclr/search?q=ApplyEditAndContinue.

So exposing the ApplyDelta metadata API for the standalone Reflection.Emit implementation should do the trick here.

**Comment by masonwheeler** [comment]
> Large chunks of Reflection.Emit code in the runtime are Windows-specific. It is not easy to just enable them for CoreCLR because of it would not work on Unix. In particular, the support for emitting debug information depends on unmanaged PDB writer that is very complex Windows-specific component that we have no plans to open source and bring x-plat.

The existence of cross-platform Roslyn suggests that this isn't a particularly intractable problem, at a high level at least.

Could this please be made a high priority?  The inability to use Reflection.Emit is currently my #1 blocker preventing me from migrating existing work to Core, and as such it's far more urgent, from my perspective at least, than the development of new features.

**Comment by AtsushiKan** [comment]
>Out of curiosity, is there any documentation / spec at all about the data format used for EnC metadata deltas, other than the actual CoreCLR source code?

Not that I've heard of but EnC is not an area where I have any knowledge.

**Comment by stakx** [comment]
@AtsushiKan - Thanks for the quick reply. :+1: I almost expected this to be a really arcane area, but I thought I'd give it a try and ask. I'll see what I can gather together on my own.

**Comment by karelz** [comment]
@stakx EnC MetaData format was never thoroughly documented (neither publicly nor internally). It was always defined just in the implementation.

**Comment by tmat** [comment]
@stakx Curious, what do you need it for?

**Comment by stakx** [comment]
@karelz - I almost suspected so... Thanks for this info! 👍

@tmat - I have spent a lot of time with System.Reflection in the past few months, and noticed its various limitations or bugs. There are sometimes situations where you'd rather just see plain, uninterpreted metadata instead of the semi-high-level view that Reflection provides (which is neither original metadata, nor as high-level as e.g. C#).

I have also discovered System.Reflection.Metadata (SRM). I have been thinking about how nice it would be if the (Core)CLR had the necessary runtime hooks and extension points so that both Reflection and Reflection.Emit could be rebuilt *on top of* SRM. (At the same time, this would enable folks to build their own reflection / dynamic code emission libraries.) Two things appear to be missing as of today to make this happen:

* Neither the CLR nor the CoreCLR runtimes allow user code to get at the metadata stream & IL of loaded assemblies/modules (so it could be inspected using e.g. SRM).
* Neither the CLR nor the CoreCLR runtime allows altering/updating an assembly that's been loaded via `Assembly.Load(byte[])` (which is what you can use today to load an assembly generated with SRM). Perhaps if EnC was exposed in some way, updating a loaded assembly would become possible.

I realise that having such runtime hooks closely allied with System.Reflection.Metadata is probably a pipe dream at this stage. But it might allow some super-interesting scenarios. So I'm trying to learn how this part of the runtime works to better gauge how big of a project this would be.

**Comment by jkotas** [comment]
> Neither the CLR nor the CoreCLR runtimes allow user code to get at the metadata stream & IL of loaded assemblies/modules, so it could be inspected using e.g. SRM

Have you seen https://github.com/dotnet/coreclr/blob/85374ceaed177f71472cc4c23c69daf7402e5048/src/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs#L25 ?

**Comment by stakx** [comment]
@jkotas - Yes, I have, and that's a really nice extension point! But it comes with some limitations: Given taht it's a `Try...` method, it may very well fail, e.g. for dynamically generated assemblies (`AssemblyBuilder`), something that today's Reflection can deal with just fine. So perhaps Reflection couldn't be built on top of that reliably.

**Comment by vermorel** [comment]
We have finally rolled our own open-source version of `AssemblyBuilder.Save` based on `System.Reflection.Metadata` (and no further dependencies). It targets .NET Standard and works under .NET Core. It does support dynamically generated assemblies.

Check-out https://github.com/Lokad/ILPack Feedback will be most welcome.

**Comment by masonwheeler** [comment]
@vermorel Looks interesting.  Does it generate PDBs too?

**Comment by vermorel** [comment]
@masonwheeler `ILPack` does not generate PDB.

**Comment by arlm** [comment]
> We have finally rolled our own open-source version of `AssemblyBuilder.Save` based on `System.Reflection.Metadata` (and no further dependencies). It targets .NET Standard and works under .NET Core. It does support dynamically generated assemblies.
> 
> Check-out https://github.com/Lokad/ILPack Feedback will be most welcome.

I tried it on a Sigil branch of mine ( https://github.com/arlm/Sigil-vNext ) nad could not make it work because it does not install on .NET Standard, only on .NET Core solutions.

**Comment by AArnott** [comment]
I tried the Lokad/ILPack project and hit [a couple of blocking issues](https://github.com/Lokad/ILPack/issues/created_by/AArnott).

**Comment by AArnott** [comment]
We need this in VS as well, as we have several libraries that use Ref.Emit and debugging such code during development is absolutely vital. We debug it primarily by writing the dynamic assembly out to disk then using peverify and ILSpy on it.

**Comment by tmat** [comment]
@AArnott Any reason not to generate .cs files and use Roslyn to build the assemblies?

**Comment by AArnott** [comment]
Yes, a few reasons @tmat:

1. Perf
1. we generate one type at a time, on demand. We don't want to compile a distinct assembly every time we need to generate a proxy. Dynamic assemblies let us add types without creating a new assembly.
1. We absolutely need to access non-public members of assemblies our dynamic code references. C# won't let us do that, but the CLR will if we write our own IL.

**Comment by tmat** [comment]
Makes sense. I assume this is for RPC proxies? Would it make sense to pre-generate them as part of VS build?

**Comment by AArnott** [comment]
Yes, RPC proxies is the main scenario, but so is generation of serializers in MessagePack. 

Pregenerating them is something we're likely to look at, but only as an optimization. We don't want to *require* that. And anyway, pregenerating or not doesn't solve our problem that we access non-public members so C# won't do, therefore we use Ref.Emit, and therefore we need to be able to debug our output and therefore need AssemblyBuilder.Save. :)

**Comment by tmat** [comment]
The non-public access can be solved. We don't want to expose such option to csc.exe but an API can be added to support the proxy generation scenario. We have other scenarios in C# scripting where this is needed.

The benefit of requiring it would be that you have just one code path to maintain. What's the downside of requiring it? If it is a default build step in VS SDK that is enabled by default, it can "just work".

**Comment by AArnott** [comment]
I think we should discuss whether we pregenerate or JIT-generate these proxies on another thread, as IMO it's orthogonal to this issue.
